sw/device/tests/sim_dv/gpio_test.c

```diff
@@ -0,0 +1,295 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_gpio.h"
+
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/base/log.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_plic.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/irq.h"
+#include "sw/device/lib/pinmux.h"
+#include "sw/device/lib/runtime/check.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/testing/test_status.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_gpio_t gpio;
+static dif_plic_t plic;
+
+static const uint32_t kNumGpios = 32;
+static const uint32_t kGpiosMask = 0xffffffff;
+// TODO #2484: GPIO pins 16-19 are special inputs - do not touch them.
+static const uint32_t kGpiosAllowedMask = 0xfff0ffff;
+
+static const uint32_t kNumChipGpios = 16;
+static const uint32_t kChipGpiosMask = 0xffff & kGpiosAllowedMask;
+
+// These indicate the GPIO pin irq expected to fire, declared volatile since
+// they are used by the interrupt handler.
+static volatile uint32_t exp_gpio_pin_irq;
+static volatile bool exp_irq_edge;
+
+/**
+ * GPIO test - verifies the GPIO pins in the input and output directions.
+ *
+ * In the output direction, SW writes the following pattern:
+ * 1. Start with GPIOs = all zeros
+ * 2. Walk a 1 through ALL GPIOs (presented by the IP), read `data_in` with each
+ *    write to ensure correctness
+ * 3. Set all GPIOs to 0s, followed by all 1s.
+ * 4. Walk a 0 through ALL GPIOs (presented by the IP), read `data_in` with each
+ *    write to ensure correctness
+ * 5. Set all GPIOs to 1s, followed by all 0s.
+ *
+ * The correctness of the GPIO values on the chip pins is verified by the
+ * external testbench. The correctness of `data_in` is limited to the number of
+ * GPIOs exposed by the chip, so we mask the written value accordingly. In
+ * addition, some GPIOs are used for 'special' functionality - JTAG TRST_N and
+ * SRST_N. We do not touch those (via mask). This will change when the jtag_mux
+ * functionality is moved into pinmux.
+ *
+ * In the input direction, the external testbench sends the following pattern:
+ * 1. Walk a 1 in 'temperature' pattern (0001, 0011, 0111, 1111, 1110, 100, ..)
+ *
+ * Both, rising and falling edges are configured for generating an interrupt. As
+ * each pin rises or falls, the SW checks the interrupt, status and `data_in`
+ * for correctness.
+ */
+
+/**
+ * Initializes PLIC and enables all GPIO interrupts.
+ */
+static void plic_init_with_irqs(mmio_region_t base_addr, dif_plic_t *plic) {
+  LOG_INFO("Initializing the PLIC.");
+
+  CHECK(dif_plic_init(base_addr, plic) == kDifPlicOk, "dif_plic_init failed");
```

Comment:
We've actually specified now that the `Ok` variant is always `0`. How would you feel about adding a `CHECK_OK` to `check.h` that's basically just `CHECKZ` with a better built-in error? E.g.,

```
#define CHECK_OK(expr) CHECKZ(expr, #expr " was not Ok");
```

This would be in a separate change, ofc.

Reply:
Sounds good to me!